test/integration/auth/rulesreview_test.go

```diff
@@ -0,0 +1,206 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"testing"
+
+	authorizationapi "k8s.io/api/authorization/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/authentication/authenticator"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	clientset "k8s.io/client-go/kubernetes"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/kubernetes/pkg/auth/nodeidentifier"
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+const (
+	nodesGroup = "system:nodes"
+	nodeUser   = "system:node:foo"
+	userGroup  = "user:foo"
+	userUser   = "user:foo"
+)
+
+type srrAuthorizer struct {
+	identifier nodeidentifier.NodeIdentifier
+}
+
+func newNodeUserInfo() user.Info {
+	return &user.DefaultInfo{
+		Name:   nodeUser,
+		Groups: []string{nodesGroup},
+	}
+}
+
+func newNonNodeUserInfo() user.Info {
+	return &user.DefaultInfo{
+		Name:   userUser,
+		Groups: []string{userGroup},
+	}
+}
+
+func (srrAuthorizer) Authorize(ctx context.Context, a authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
+	return authorizer.DecisionAllow, "", nil
+}
+
+func (r *srrAuthorizer) RulesFor(user user.Info, namespace string) ([]authorizer.ResourceRuleInfo, []authorizer.NonResourceRuleInfo, bool, error) {
+	if _, isNode := r.identifier.NodeIdentity(user); isNode {
+		// indicate nodes do not have fully enumerated permissions
+		return nil, nil, true, fmt.Errorf("node authorizer does not support user rule resolution")
+	}
+	return nil, nil, false, nil
+}
+
+func TestSelfSubjectRulesReviewNodeUser(t *testing.T) {
+	masterConfig := framework.NewIntegrationTestMasterConfig()
+	masterConfig.GenericConfig.Authentication.Authenticator = authenticator.RequestFunc(func(req *http.Request) (*authenticator.Response, bool, error) {
+		return &authenticator.Response{
+			User: newNodeUserInfo(),
+		}, true, nil
+	})
+	masterConfig.GenericConfig.Authorization.Authorizer = srrAuthorizer{}
+	_, s, closeFn := framework.RunAMaster(masterConfig)
+	defer closeFn()
+
+	clientset := clientset.NewForConfigOrDie(&restclient.Config{Host: s.URL})
+
+	tests := []struct {
+		name           string
+		username       string
```

Comment:
Needs to be `user.Info` that needs to be impersonated.

Reply:
updated as suggested